squint.c

```diff
@@ -0,0 +1,1611 @@
+            if ( ((mode == 0) && !hasLink) ||   // intra-function mode
+                 ((mode != 0) &&  hasLink) ) {  // inter-function mode
+               /* add branch to table */
+               permutation[branchCount] = branchCount;
+               branchAddr[branchCount] = (scan - funcBegin);
+               tmp = (*scan & 0x00ffffff) |
+                     ((*scan & 0x00800000) ? 0xff000000 : 0);
+               branchTarget[branchCount] = (scan - funcBegin) + 2 + tmp;
+               ++branchCount;
+            }
+         }
+      }
+
+      /* sort in terms of target address */
+      /* this is an order preserving sort */
+      /* for secondary keys */
+      ii = 0;
+      do {
+         done = 1;
+         for (jj = branchCount-2; jj >= ii; --jj) {
+            if (branchTarget[jj+1] < branchTarget[jj]) {
+               tmp = branchTarget[jj+1];
+               branchTarget[jj+1] = branchTarget[jj];
+               branchTarget[jj] = tmp;
+               tmp = permutation[jj+1];
+               permutation[jj+1] = permutation[jj];
+               permutation[jj] = tmp;
+               done = 0;
+            }
+         }
+         ++ii;
+      } while (!done);
+   }
+
+   if (nopCount > 0) { // there are nops to remove
+
+      currAddr = 0;
+      currTarget = 0;
+      offset = 0;
+
+      packed = funcBegin;
+      for (scan = funcBegin; scan <= funcEnd; ++scan, ++offset) {
+         align = 1;
+         while (currTarget < branchCount) {
+            if (branchTarget[currTarget] == offset) {
+               if (align && ((mode != 0) ||
+                   branchAddr[permutation[currTarget]] > offset)) {
+                  /* quadword align loop branch target */
+                  tmp = 4 - ((packed-funcBegin) & 3);
+                  if (tmp != 4 && (scan-packed) >= tmp) {
+                     while (tmp-- > 0)
+                        *packed++ = NOP;
+                  }
+                  align = 0;
+               }
+               branchTarget[currTarget++] = packed - funcBegin;
+            }
+            else {
+               break;
+            }
+         }
+         if (currAddr < branchCount) {
+            if (branchAddr[currAddr] == offset) {
+               branchAddr[currAddr++] = packed - funcBegin;
+            }
+         }
+         if (is_const(scan)) {
+            rel_pc_const(packed, scan);
+            ++packed;
+         }
+         else if (*scan != NOP) {
+            if ((*scan>>16 & 0x0fff) == 0x59f) { // ldr rx, [pc, #x]
+               rel_pc_ldr(packed, scan);
+               ++packed;
+               /* remap pc-relative load operations */
+               // *scan += (scan - packed)*sizeof(int);
+            }
+            else if ((*scan & 0x0e000000) == 0x0a000000) {
+               if (*scan & (1<<24)) {
+                  // adjust bl instruction address
+                  // jk this is bogus since for mode 0 only
+                  *scan += scan - packed;
```

Comment:
Shouldn't we check the direction of the branch before adding the difference?

If the destination is less than the current address, we should subtract the difference:

```
func1:
  xx
  xx

func2:
  xx
  xx
  nop
  nop
  bl func1
```

If the destination is greater than the current address, we should add the difference:


```
func1:
  xx
  xx
  nop
  nop
  bl func2

func2:
  xx
  xx
```

Reply:
Nope. This code is solid, and the correction is made elsewhere.  As my comment in this code fragment implies, there is a bug  here (the one I indicated near the top of this pull request thread) that does not have an effect because the inter-procedural optimization is never called in Squint.